{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b9ea1a80-b951-4457-8815-075ae99bd4ba",
   "metadata": {},
   "source": [
    "# Working with h5ad files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b95bf0a-2613-4900-a271-da5a41fd7e5e",
   "metadata": {},
   "source": [
    "This notebook will discuss how cells are linked to the cell-by-gene data stored in .h5ad files. It will also discuss how to work with .h5ad files using the anndata library."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b57d216c-44fc-4249-a2c0-9455a58276f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import anndata\n",
    "import h5py\n",
    "import numpy as np\n",
    "import os\n",
    "import pandas as pd\n",
    "import pathlib\n",
    "import scipy.sparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a10bc4fd-cf22-4d69-bb3e-8afa4b414991",
   "metadata": {},
   "outputs": [],
   "source": [
    "import abc_atlas_access.abc_atlas_cache.abc_project_cache as abc_project_cache"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9aac7b97-791d-41a5-bec6-598dbcf4319d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/scott.daniel/miniconda3/envs/webinar_2026_02/lib/python3.12/site-packages/abc_atlas_access/abc_atlas_cache/cloud_cache.py:679: OutdatedManifestWarning: You are loading releases/20250531/manifest.json. A more up to date version of the dataset -- releases/20251031/manifest.json -- exists online. To see the changes between the two versions of the dataset, run\n",
      "type.compare_manifests('releases/20250531/manifest.json', 'releases/20251031/manifest.json')\n",
      "To load another version of the dataset, run\n",
      "type.load_manifest('releases/20251031/manifest.json')\n",
      "  warnings.warn(msg, OutdatedManifestWarning)\n"
     ]
    }
   ],
   "source": [
    "cache_dir = \"/Users/scott.daniel/KnowledgeEngineering/cell_type_mapper/examples/data/abc_cache\"\n",
    "\n",
    "if not os.path.isdir(cache_dir):\n",
    "    raise RuntimeError(\n",
    "        \"Set cache_dir above to the path to the directory where you want to download ABC Atlas data\"\n",
    "    )\n",
    "\n",
    "abc_cache = abc_project_cache.AbcProjectCache.from_cache_dir(cache_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e3f4232-5e2e-4832-94c2-a43329330a2f",
   "metadata": {},
   "source": [
    "# 1. What are h5ad files?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d88c0090-459d-4cb1-978f-4bad12d3ab11",
   "metadata": {},
   "source": [
    "h5ad files are how the actual cell-by-gene expression data is distributed in `abc_atlas_access`. They are a file format designed to work with the Python library anndata.\n",
    "\n",
    "anndata is a Python library and framework for manipulating data and metadata simultaneously (the \"ann\" stands for \"annotated\"). [Their official documentation is here](https://anndata.readthedocs.io/en/stable/). h5ad is a file format, based on HDF5, for serializing and distributing anndata objects. All of the cell-by-gene data in `abc_atlas_access` is distributed as .h5ad files. h5ad files are [documented here](https://anndata.readthedocs.io/en/stable/fileformat-prose.html)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb338835-1e8d-427e-ad07-f842b4dc8158",
   "metadata": {},
   "source": [
    "## Create cartoon data for examples\n",
    "\n",
    "Please just run the cells in this section without worrying about what they do. The actual data in `abc_atlas_access` is very large. In order to demonstrate some of the subtleties involved in working with `h5ad` files without crashing your computer, we are going to generate some smaller cartoon data here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3116ad81-04ac-48e6-8136-1101c8dbb665",
   "metadata": {},
   "outputs": [],
   "source": [
    "cartoon_data_dir = pathlib.Path('cartoon_data')\n",
    "\n",
    "if not cartoon_data_dir.exists():\n",
    "    cartoon_data_dir.mkdir()\n",
    "\n",
    "if not cartoon_data_dir.is_dir():\n",
    "    raise RuntimeError(f'{cartoon_data_dir} is not a dir')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0e3fbcea-688e-4a58-ab1e-6e5e459e7e8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "n_cells = 300000\n",
    "n_genes = 5000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "051ec871-e990-48c0-a843-e24d815e4f3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_data(dst_dir, n_cells, n_genes):\n",
    "    csr_path = dst_dir / 'as_csr.h5ad'\n",
    "    csc_path = dst_dir / 'as_csc.h5ad'\n",
    "    dense_path = dst_dir / 'as_dense.h5ad'\n",
    "\n",
    "    rng = np.random.default_rng(6713121)\n",
    "\n",
    "    obs = pd.DataFrame(\n",
    "        [{'cell': f'cell_{ii}', 'donor': f'donor_{rng.integers(0, 3)}'}\n",
    "         for ii in range(n_cells)]\n",
    "    ).set_index('cell')\n",
    "\n",
    "    var = pd.DataFrame(\n",
    "        [{'gene_id': f'ENS{ii:07d}', 'symbol': f'sym{ii}'}\n",
    "         for ii in range(n_genes)]\n",
    "    ).set_index('gene_id')\n",
    "\n",
    "\n",
    "    data = np.zeros(n_cells*n_genes, dtype=np.int32)\n",
    "    chunk_size = n_cells*n_genes//5\n",
    "    for i0 in range(0, n_cells*n_genes, chunk_size):\n",
    "        i1 = min(n_cells*n_genes, i0+chunk_size)\n",
    "        chosen_idx = rng.choice(np.arange(i0, i1, 1), (i1-i0)//5, replace=False)\n",
    "        data[chosen_idx] = rng.integers(1, 999, len(chosen_idx)).astype(np.int32)\n",
    "    data = data.reshape((n_cells, n_genes))\n",
    "    print('created data')\n",
    "\n",
    "    as_dense = anndata.AnnData(obs=obs, var=var, X=data)\n",
    "    as_dense.write_h5ad(dense_path)\n",
    "    del as_dense\n",
    "    print(f'wrote {dense_path}')\n",
    "\n",
    "    as_csc = anndata.AnnData(obs=obs, var=var, X=scipy.sparse.csc_matrix(data))\n",
    "    as_csc.write_h5ad(csc_path)\n",
    "    del as_csc\n",
    "    print(f'wrote {csc_path}')\n",
    "\n",
    "    as_csr = anndata.AnnData(obs=obs, var=var, X=scipy.sparse.csr_matrix(data))\n",
    "    as_csr.write_h5ad(csr_path)\n",
    "    del as_csr\n",
    "    print(f'wrote {csr_path}')\n",
    "\n",
    "    return {\n",
    "        'dense': dense_path,\n",
    "        'csc_path': csc_path,\n",
    "        'csr_path': csr_path\n",
    "    }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "95f88048-f4b1-40f8-a5b8-f23a9b776f63",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "created data\n",
      "wrote cartoon_data/as_dense.h5ad\n",
      "wrote cartoon_data/as_csc.h5ad\n",
      "wrote cartoon_data/as_csr.h5ad\n"
     ]
    }
   ],
   "source": [
    "path_lookup = create_data(\n",
    "    dst_dir=cartoon_data_dir,\n",
    "    n_cells=n_cells,\n",
    "    n_genes=n_genes\n",
    ")\n",
    "\n",
    "dense_path = path_lookup['dense']\n",
    "csc_path = path_lookup['csc_path']\n",
    "csr_path = path_lookup['csr_path']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f91d38e-c410-48cd-9373-2ad0711b7a4f",
   "metadata": {},
   "source": [
    "## Anatomy of h5ad files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bedd372d-f948-49c3-838b-509ac9b16fb3",
   "metadata": {},
   "source": [
    "h5ad files are serializations of `AnnData` objects. `AnnData` objects are designed to hold cell-by-gene expression data alongside a lot of optional metadata. The h5ad files distributed by `abc_atlas_access` are relatively stripped down. We do not exercise all of the options that the anndata library makes available. In this notebook, we will discuss the features of these stripped down `AnnData` objects. To see all of the features available, consult [the official anndata documentation](https://anndata.readthedocs.io/en/stable/).\n",
    "\n",
    "First, let us open one of our example h5ad files.\n",
    "\n",
    "**Note:** when dealing with real (large) h5ad files, the `backed='r'` kwarg is very important. Without that, anndata will try to read all of the data in the h5ad file into memory at once, which could be prohibitively large. Reading the file with `backed='r'` as below gives you the option to only read data into memory when you need it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "12cc0138-6ad2-41ab-aa17-91cac2c460ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ad_example = anndata.read_h5ad(dense_path, backed='r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d494f3a8-f6df-4ab9-a51c-dde7fb26e444",
   "metadata": {},
   "source": [
    "`AnnData` objects are arranged around arrays referred to as `X` in which rows are cells and columns are genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7374190c-cdc0-41b9-890a-81afa7066c73",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<HDF5 dataset \"X\": shape (300000, 5000), type \"<i4\">"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5ad_example.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50bd1503-a3be-4033-b904-30a3f3e962a8",
   "metadata": {},
   "source": [
    "Here we have a dataset with 300000 cells and 5000 genes (which, consulting the cells in which we created this cartoon data, is what we expected).\n",
    "\n",
    "The `AnnData` object also carries with it two pandas DataFrames.\n",
    "- `obs` contains all of the metadata related to the cells in the dataset\n",
    "- `var` contains all of the metadata related to the genes in the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d056f96b-f2a4-4c6e-bc28-899799a893ee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cell_0</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_1</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_2</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_3</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_4</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299995</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299996</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299997</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299998</th>\n",
       "      <td>donor_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299999</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300000 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               donor\n",
       "cell                \n",
       "cell_0       donor_1\n",
       "cell_1       donor_1\n",
       "cell_2       donor_1\n",
       "cell_3       donor_0\n",
       "cell_4       donor_1\n",
       "...              ...\n",
       "cell_299995  donor_0\n",
       "cell_299996  donor_0\n",
       "cell_299997  donor_1\n",
       "cell_299998  donor_2\n",
       "cell_299999  donor_0\n",
       "\n",
       "[300000 rows x 1 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5ad_example.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d1735cdf-e9c0-4c94-96c4-c3e4cadb1d5f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENS0000000</th>\n",
       "      <td>sym0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000001</th>\n",
       "      <td>sym1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000002</th>\n",
       "      <td>sym2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000003</th>\n",
       "      <td>sym3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000004</th>\n",
       "      <td>sym4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004995</th>\n",
       "      <td>sym4995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004996</th>\n",
       "      <td>sym4996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004997</th>\n",
       "      <td>sym4997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004998</th>\n",
       "      <td>sym4998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004999</th>\n",
       "      <td>sym4999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             symbol\n",
       "gene_id            \n",
       "ENS0000000     sym0\n",
       "ENS0000001     sym1\n",
       "ENS0000002     sym2\n",
       "ENS0000003     sym3\n",
       "ENS0000004     sym4\n",
       "...             ...\n",
       "ENS0004995  sym4995\n",
       "ENS0004996  sym4996\n",
       "ENS0004997  sym4997\n",
       "ENS0004998  sym4998\n",
       "ENS0004999  sym4999\n",
       "\n",
       "[5000 rows x 1 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5ad_example.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68666dc0-724d-4584-8c0e-e6e6225d0a76",
   "metadata": {},
   "source": [
    "To get `X` as a numpy array, call `to_memory()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "4f91c2ec-0b82-4739-a207-f1a645ae0d5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ad_example = h5ad_example.to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6562fba9-6d13-43cd-a535-1dee36ddd426",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  0,   0,   0, ...,   0,   0,   0],\n",
       "       [  0,   0,   0, ...,   0,   0,   0],\n",
       "       [  0,   0,   0, ...,   0,   0,   0],\n",
       "       ...,\n",
       "       [759,   0,   0, ...,   0,   0, 644],\n",
       "       [  0,   0,   0, ...,   0,   0,   0],\n",
       "       [712, 947,   0, ...,   0,   0, 853]],\n",
       "      shape=(300000, 5000), dtype=int32)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5ad_example.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04833096-c1fc-4c9a-9ee4-4aaba84395e3",
   "metadata": {},
   "source": [
    "`X` is now a numpy array that can be manipulated like any other numpy array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0bebffef-99ba-478e-8ceb-b175c8b52332",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([491524, 467316, 526166, ..., 490179, 488071, 524453],\n",
       "      shape=(300000,))"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5ad_example.X.sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "8c79ea2c-c40e-411d-978e-e86630a1bdb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 99.50587   , 100.44607333,  99.50548   , ...,  99.70915   ,\n",
       "        99.82104333, 100.55265   ], shape=(5000,))"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5ad_example.X.mean(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1edb6fc-a110-4521-81a7-485646763db4",
   "metadata": {},
   "source": [
    "**Note:** the anndata library is not always as careful as we would like about closing file handles. When deleting an `AnnData` object you have read from disk, it behoves you to call `anndata_obj.file.close()` to make sure the file you read it from is not left open."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "23fbe543-d851-4a19-85c4-4490bc724ab3",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ad_example.file.close()\n",
    "del h5ad_example"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba6b0c1e-dac4-4b43-bbdd-0e8c1d2c71b7",
   "metadata": {},
   "source": [
    "Often, the cell-by-gene data stored in the `AnnData` object is very sparse. It is efficient, in these cases, to store the data as a sparse array. This is what we did in the `csr_path` and `csc_path` files above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d190bf17-421d-445d-89f7-b2d11ae4adb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ad_example = anndata.read_h5ad(csr_path, backed='r')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "49743700-64e8-464f-b217-ae887329f186",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CSRDataset: backend hdf5, shape (300000, 5000), data_dtype int32"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5ad_example.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "48536ea1-a436-437c-96e2-36fdbf3722de",
   "metadata": {},
   "outputs": [],
   "source": [
    "h5ad_example = h5ad_example.to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "fb28ff2c-e3c8-4598-9ea2-6a8cf8929585",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Compressed Sparse Row sparse matrix of dtype 'int32'\n",
       "\twith 300000000 stored elements and shape (300000, 5000)>"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h5ad_example.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f7ae8c9-848f-458b-87c2-2e90a02b3db2",
   "metadata": {},
   "source": [
    "In this case, `X` is stored as a sparse matrix as defined by [the scipy.sparse module](https://docs.scipy.org/doc/scipy/reference/sparse.html). Many numerical operations that can be performed on numpy arrays can also be performed on these sparse matrices (albeit somewhat slower)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "13b3a8c2-ca6c-4227-a891-68e4046c1942",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 201 ms, sys: 797 ms, total: 997 ms\n",
      "Wall time: 1.05 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "matrix([[491524],\n",
       "        [467316],\n",
       "        [526166],\n",
       "        ...,\n",
       "        [490179],\n",
       "        [488071],\n",
       "        [524453]], shape=(300000, 1))"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "h5ad_example.X.sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "20eeb7e4-994d-4d73-8c78-9c3ce7453c60",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 711 ms, sys: 1.41 s, total: 2.12 s\n",
      "Wall time: 2.3 s\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "matrix([[ 99.50587   , 100.44607333,  99.50548   , ...,  99.70915   ,\n",
       "          99.82104333, 100.55265   ]], shape=(1, 5000))"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "h5ad_example.X.mean(axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86566ed7-b9cf-4eae-868d-771aba017eb4",
   "metadata": {},
   "source": [
    "To convert the arrays into dense (with all of the zeros included) numpy arrays, call `toarray()` (this will, of course, incur a memory penalty as a lot more data will be stored in RAM)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2135c116-3eda-4038-94b1-6e5fb9214e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "dense_x = h5ad_example.X.toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2eda4d7e-3d0a-4c03-b817-ba7cea04e9ca",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 288 ms, sys: 19.5 ms, total: 308 ms\n",
      "Wall time: 307 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([491524, 467316, 526166, ..., 490179, 488071, 524453],\n",
       "      shape=(300000,))"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "dense_x.sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "55f693ef-b1fe-4739-a743-e3ee373701b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 643 ms, sys: 21.3 ms, total: 665 ms\n",
      "Wall time: 667 ms\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([ 99.50587   , 100.44607333,  99.50548   , ...,  99.70915   ,\n",
       "        99.82104333, 100.55265   ], shape=(5000,))"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%%time\n",
    "dense_x.mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "67db4d06-77e8-4599-ae8d-316f2464d7c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "del dense_x\n",
    "h5ad_example.file.close()\n",
    "del h5ad_example"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "161fde0f-7835-4ae9-8d55-222fba489bad",
   "metadata": {},
   "source": [
    "## Slicing h5ad files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "367c2972-cff6-4c5c-9d8b-91ef7003c611",
   "metadata": {},
   "source": [
    "Often, users want to subset cell-by-gene data either by cells or by genes. In this section, we discuss now to perform this operation for `AnnData` objects, and what the flavor of storage (dense versus sparse) means about various slicing operations.\n",
    "\n",
    "There are two flavors of sparse matrix that anndata supports\n",
    "- CSR (\"Compressed Sparse Row\") matrices are stored in such a way that they are optimized for slicing by rows (cells)\n",
    "- CSC (\"Compressed Sparse Column\") matrices are stored in such a way that they are optimized for slicing by columns (genes)\n",
    "\n",
    "In the cells above, we wrote the same cell-by-gene data three ways:\n",
    "- `dense` data is stored as an (n_cells, n_genes) array with all elements, (even those that are zero) saved to disk\n",
    "- `csr` data is stored in \"Compressed Sparse Row\" format, which is optimized for slicing on rows (cells)\n",
    "- `csc` data is stored in \"Compressed Sparse Column\" format, which is optimized for slcing on columns (genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "eda0f499-cc36-4f8c-b8cc-fc9fb6b30a28",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "as_dense.h5ad is 5.60e+00 GB in size\n",
      "as_csr.h5ad is 2.25e+00 GB in size\n",
      "as_csc.h5ad is 2.25e+00 GB in size\n"
     ]
    }
   ],
   "source": [
    "for file_path in (dense_path, csr_path, csc_path):\n",
    "    gb = os.stat(file_path).st_size / (1024**3)\n",
    "    print(f'{file_path.name} is {gb:.2e} GB in size')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4c8950e5-6750-4be0-9de8-5426b25b99ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "dense = anndata.read_h5ad(dense_path, backed='r')\n",
    "csc = anndata.read_h5ad(csc_path, backed='r')\n",
    "csr = anndata.read_h5ad(csr_path, backed='r')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b1a3946-12ab-452b-beb3-bcec8d72b921",
   "metadata": {},
   "source": [
    "Not the different representations of `X`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b6646d76-712a-44be-9f84-bffb2d067385",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<HDF5 dataset \"X\": shape (300000, 5000), type \"<i4\">"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dense.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "812805a7-3867-41b4-819d-64cc44931afd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CSCDataset: backend hdf5, shape (300000, 5000), data_dtype int32"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csc.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "70880d1d-799c-461b-bb9b-9df47f1076b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CSRDataset: backend hdf5, shape (300000, 5000), data_dtype int32"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "csr.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7ee3c69-c4b7-4348-885d-e8d9c2384f59",
   "metadata": {},
   "source": [
    "### Slicing by cells"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "511d2f23-9273-49e1-b514-53f482a1b2a5",
   "metadata": {},
   "source": [
    "Ultimately, you can slice an anndata object the same way you would slice a numpy array. For instance, to grab a specific set of cells denoted by the indices (i.e. their row numbers) you could just:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "560171e9-c9b3-42bf-8615-9789032fc384",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_idx = np.array([0, 3, 5, 290000, 290001, 290002])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4bfb762c-b690-45d8-9fb9-3ceda2410698",
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_of_cells = dense[cell_idx, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c288ecd0-aba9-4d4d-bf6a-718616fb3859",
   "metadata": {},
   "source": [
    "You now have an `AnnData` object sliced to include only the cells specified by `cell_idx`. Note that the slicing is automatically applied to both `X` and `obs`. `var` is unaffected since we did not slice on genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7a1660ac-0074-4a2f-910d-79b6db56350d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  0,   0,   0, ...,   0,   0,   0],\n",
       "       [  0, 575,   0, ...,   0,   0, 501],\n",
       "       [899,   0,   0, ...,   0,   0,  70],\n",
       "       [  0,   0,   0, ...,   0,   0,   0],\n",
       "       [  0,   0,   0, ..., 330,   0,   0],\n",
       "       [  0, 821,   0, ...,   0,   0,   0]], shape=(6, 5000), dtype=int32)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_of_cells.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "b07e7233-e914-4634-ae0d-e560458d1af1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cell_0</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_3</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_5</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_290000</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_290001</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_290002</th>\n",
       "      <td>donor_2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               donor\n",
       "cell                \n",
       "cell_0       donor_1\n",
       "cell_3       donor_0\n",
       "cell_5       donor_1\n",
       "cell_290000  donor_1\n",
       "cell_290001  donor_0\n",
       "cell_290002  donor_2"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_of_cells.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "42c48d84-36ae-4631-8ff0-32b69fa8c156",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENS0000000</th>\n",
       "      <td>sym0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000001</th>\n",
       "      <td>sym1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000002</th>\n",
       "      <td>sym2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000003</th>\n",
       "      <td>sym3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000004</th>\n",
       "      <td>sym4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004995</th>\n",
       "      <td>sym4995</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004996</th>\n",
       "      <td>sym4996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004997</th>\n",
       "      <td>sym4997</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004998</th>\n",
       "      <td>sym4998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0004999</th>\n",
       "      <td>sym4999</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5000 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             symbol\n",
       "gene_id            \n",
       "ENS0000000     sym0\n",
       "ENS0000001     sym1\n",
       "ENS0000002     sym2\n",
       "ENS0000003     sym3\n",
       "ENS0000004     sym4\n",
       "...             ...\n",
       "ENS0004995  sym4995\n",
       "ENS0004996  sym4996\n",
       "ENS0004997  sym4997\n",
       "ENS0004998  sym4998\n",
       "ENS0004999  sym4999\n",
       "\n",
       "[5000 rows x 1 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_of_cells.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f06db4c0-eac3-4424-abbb-fbe2c61339e0",
   "metadata": {},
   "source": [
    "However, slicing the three `AnnData` objects (which contain nominally identical data) along the cell axis will take very different resources.\n",
    "\n",
    "Let us subset our three `AnnData` objects along the cell axis, using Jupyter's `%%time` magic function to see how long each operation takes.\n",
    "\n",
    "**Note:** the `to_memory()` call is necessary to tell `anndata` to actually load the subset of data into memory. Otherwise, `anndata` will just construct a pointer to the subset of cells you need, and everything will appear lightning fast until you actually try to process the subsetted data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "87c5eb76-f8f4-426c-ba35-710fa0a246a5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.82 ms, sys: 763 μs, total: 2.58 ms\n",
      "Wall time: 1.96 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from_dense = dense[cell_idx, :].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "b3c80c90-6d02-41b2-ab68-a815b42fe6bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 2.04 ms, sys: 1.49 ms, total: 3.53 ms\n",
      "Wall time: 2.81 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from_csr = csr[cell_idx, :].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fd0feac6-7476-44a4-9be0-3752b520c61a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 329 ms, sys: 457 ms, total: 786 ms\n",
      "Wall time: 1.06 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from_csc = csc[cell_idx, :].to_memory()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "379b2349-9608-486c-bdf1-6ed3b5704526",
   "metadata": {},
   "source": [
    "Note how much longer it took to subset the CSC `AnnData` object than either the dense or the CSR `AnnData` object. Nevertheless, the three operations produced the same output."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "105822f4-b9d5-4337-bd45-ef2ebc986bb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.testing.assert_array_equal(from_dense.X, from_csc.X.toarray())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "a6cb93b1-0f47-47c1-86d1-f3be1d68e2d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.testing.assert_array_equal(from_dense.X, from_csr.X.toarray())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36595ad0-df7c-475f-b68e-5c4a751cc1b5",
   "metadata": {},
   "source": [
    "To see why this is the case, let's call the convenience function `read_and_profile.read` defined in the `read_and_profile.py` code that came with this notebook. This function spins up a child process to subset the h5ad file, using the parent process to keep track of the maximum amount of memory used by the reading process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "f5c3407e-48ce-43e0-9c0d-d0a93a28edbc",
   "metadata": {},
   "outputs": [],
   "source": [
    "import read_and_profile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6c6233c5-0095-4754-9876-0312802c1d24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slicing as_dense.h5ad along 'cell'\n",
      "took 2.49e+00 ms\n",
      "maximum memory footprint 1.58e-01 GB (file size is 5.60e+00 GB)\n"
     ]
    }
   ],
   "source": [
    "read_and_profile.read(\n",
    "    path=dense_path,\n",
    "    axis='cell',\n",
    "    idx=cell_idx\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a18f4a2a-9e97-4c2f-9dbd-e98897327b36",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slicing as_csr.h5ad along 'cell'\n",
      "took 3.45e+00 ms\n",
      "maximum memory footprint 1.58e-01 GB (file size is 2.25e+00 GB)\n"
     ]
    }
   ],
   "source": [
    "read_and_profile.read(\n",
    "    path=csr_path,\n",
    "    axis='cell',\n",
    "    idx=cell_idx\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "47394990-89b8-4cc5-aaed-39649da8e539",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slicing as_csc.h5ad along 'cell'\n",
      "took 7.18e+02 ms\n",
      "maximum memory footprint 2.44e+00 GB (file size is 2.25e+00 GB)\n"
     ]
    }
   ],
   "source": [
    "read_and_profile.read(\n",
    "    path=csc_path,\n",
    "    axis='cell',\n",
    "    idx=cell_idx\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff5729e7-7da6-411e-a033-efba17bb5e49",
   "metadata": {},
   "source": [
    "The salient feature to notice is that the memory footprint of the processes which subset the dense and CSR matrices along the cell axis are significantly smaller than the size of the file. However, in order to subset the CSC matrix along the cell axis, the entire file had to be loaded into memory. We see something similar when we slice along the gene axis: the dense and CSC matrices can be efficiently sliced along the gene axis. The CSR matrix cannot."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ef15dcc-a009-4b73-86c8-010d93576617",
   "metadata": {},
   "source": [
    "## Slicing by genes\n",
    "\n",
    "If we subset by genes, we will see that the opposite is true. Subsetting the CSR `AnnData` object will take much longer than subsetting either the dense or the CSC `AnnData` object since the CSR `AnnData` object is optimized for subsetting on cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "baebe264-b045-4943-8b4b-83df92f67e25",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_idx = np.array([5, 6, 7, 1501, 1502, 1503])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "51f9945f-c0c5-4388-afef-99cd97fab033",
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_of_genes = dense[:, gene_idx]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "1ad6308c-3698-45f5-84ae-ff36acc38f88",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[  0, 524,   0,   0,   0,   0],\n",
       "       [  0,   0, 679,   0,   0,   0],\n",
       "       [  0, 907,   0,   0,  70,   0],\n",
       "       ...,\n",
       "       [  0,   0, 192,   0,   0,   0],\n",
       "       [  0,   0,   0, 318,   0,   0],\n",
       "       [  0,   0,   0,   0,   0, 370]], shape=(300000, 6), dtype=int32)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_of_genes.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "36edea4a-0711-4f52-8407-5c7b36ca656d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cell_0</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_1</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_2</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_3</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_4</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299995</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299996</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299997</th>\n",
       "      <td>donor_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299998</th>\n",
       "      <td>donor_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_299999</th>\n",
       "      <td>donor_0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>300000 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               donor\n",
       "cell                \n",
       "cell_0       donor_1\n",
       "cell_1       donor_1\n",
       "cell_2       donor_1\n",
       "cell_3       donor_0\n",
       "cell_4       donor_1\n",
       "...              ...\n",
       "cell_299995  donor_0\n",
       "cell_299996  donor_0\n",
       "cell_299997  donor_1\n",
       "cell_299998  donor_2\n",
       "cell_299999  donor_0\n",
       "\n",
       "[300000 rows x 1 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_of_genes.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f4aac509-7ac5-4368-89c2-25afdb83b401",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENS0000005</th>\n",
       "      <td>sym5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000006</th>\n",
       "      <td>sym6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0000007</th>\n",
       "      <td>sym7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0001501</th>\n",
       "      <td>sym1501</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0001502</th>\n",
       "      <td>sym1502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENS0001503</th>\n",
       "      <td>sym1503</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             symbol\n",
       "gene_id            \n",
       "ENS0000005     sym5\n",
       "ENS0000006     sym6\n",
       "ENS0000007     sym7\n",
       "ENS0001501  sym1501\n",
       "ENS0001502  sym1502\n",
       "ENS0001503  sym1503"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subset_of_genes.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "411a9cd1-2c57-48be-a050-4eba4487cc0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 42.5 ms, sys: 432 ms, total: 475 ms\n",
      "Wall time: 475 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from_dense = dense[:, gene_idx].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "961766af-98af-4edc-8b73-757151145752",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 361 ms, sys: 513 ms, total: 874 ms\n",
      "Wall time: 1.19 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from_csr = csr[:, gene_idx].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "10ee48a9-9ad5-4868-930a-79732db07101",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 28.4 ms, sys: 6.14 ms, total: 34.5 ms\n",
      "Wall time: 35.8 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "from_csc = csc[:, gene_idx].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5ba36d42-f0fc-4f31-9d1c-4f67ce49eb9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.testing.assert_array_equal(from_dense.X, from_csc.X.toarray())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "641b72d6-766b-452b-8f11-3b7ac428f0e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.testing.assert_array_equal(from_dense.X, from_csr.X.toarray())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "06484f59-a244-4c4e-a975-ebf33ad808c6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slicing as_dense.h5ad along 'gene'\n",
      "took 5.30e+02 ms\n",
      "maximum memory footprint 1.72e-01 GB (file size is 5.60e+00 GB)\n"
     ]
    }
   ],
   "source": [
    "read_and_profile.read(\n",
    "    path=dense_path,\n",
    "    axis='gene',\n",
    "    idx=gene_idx\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "f2f8be14-47f1-4559-849d-d8d423739d74",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slicing as_csr.h5ad along 'gene'\n",
      "took 1.12e+03 ms\n",
      "maximum memory footprint 2.43e+00 GB (file size is 2.25e+00 GB)\n"
     ]
    }
   ],
   "source": [
    "read_and_profile.read(\n",
    "    path=csr_path,\n",
    "    axis='gene',\n",
    "    idx=gene_idx\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "cc919832-7d3a-499f-b52b-d9cf53fcfa43",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slicing as_csc.h5ad along 'gene'\n",
      "took 3.16e+01 ms\n",
      "maximum memory footprint 1.70e-01 GB (file size is 2.25e+00 GB)\n"
     ]
    }
   ],
   "source": [
    "read_and_profile.read(\n",
    "    path=csc_path,\n",
    "    axis='gene',\n",
    "    idx=gene_idx\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e2e1b94-a3af-4474-9978-9ee3bfafb6f7",
   "metadata": {},
   "source": [
    "Again, subsetting the CSR `AnnData` object along its sub-optimal axis involves reading the entire file into memory. This could be a problem when dealing with actual cell-by-gene data, which tends to be quite large."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b20b80c8-f2bf-4d2b-b634-ea48202d6f56",
   "metadata": {},
   "source": [
    "### Takeaway\n",
    "\n",
    "Before deciding how to manipulate an AnnData object, check to see how the data is stored on disk. You can probably infer this just by printing the result of `my_anndata.X` to stdout, i.e."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "711d408f-2a2b-4f4c-997e-e2219ce86718",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "as_dense.h5ad -- <HDF5 dataset \"X\": shape (300000, 5000), type \"<i4\">\n",
      "as_csr.h5ad -- CSRDataset: backend hdf5, shape (300000, 5000), data_dtype int32\n",
      "as_csr.h5ad -- CSCDataset: backend hdf5, shape (300000, 5000), data_dtype int32\n"
     ]
    }
   ],
   "source": [
    "print(dense_path.name, '--', dense.X)\n",
    "print(csr_path.name, '--', csr.X)\n",
    "print(csr_path.name, '--', csc.X)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddd1624c-613c-44ff-a88d-cb2c87d90538",
   "metadata": {},
   "source": [
    "# 2. Actual abc_atlas_access data\n",
    "\n",
    "Now it is time for us to work with actual h5ad files distributed with `abc_atlas_access`."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfe1bb11-e3ad-42ce-a427-65e3e6dcdaf5",
   "metadata": {},
   "source": [
    "## How are cells linked to h5ad files -- the cell_metadata\n",
    "\n",
    "The metadata for all of the cells in any given ABC Atlas dataset is in the the `cell_metadata` .csv file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "c86a9546-21f4-4f0d-809a-91b0fd523907",
   "metadata": {},
   "outputs": [],
   "source": [
    "wmb_metadata = abc_cache.get_metadata_dataframe(\n",
    "    directory='WMB-10X',\n",
    "    file_name='cell_metadata'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9aed448-af58-4e4e-8ce9-b2fdd1f0a95b",
   "metadata": {},
   "source": [
    "The linkage to h5ad files is encoded in the\n",
    "- `dataset_label` column, which refers to the `abc_atlas_access` directory where the cell-by-gene data for a cell can be found\n",
    "- `feature_matrix_label` column, which refers to the file name within that directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "9ae24fc0-2fcf-435c-a1da-4b31a98a6372",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_label</th>\n",
       "      <th>dataset_label</th>\n",
       "      <th>feature_matrix_label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1889336</th>\n",
       "      <td>GTTACAGCAGCAATTC-294_A05</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>WMB-10Xv3-TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1960181</th>\n",
       "      <td>AATCACGAGTGGATTA-448_A07</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>WMB-10Xv3-Isocortex-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3813712</th>\n",
       "      <td>GTGCATAGTCTGGTCG-069_C01</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>WMB-10Xv2-Isocortex-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3742959</th>\n",
       "      <td>TTCTTAGAGAACAATC-092_A01</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>WMB-10Xv2-OLF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3407039</th>\n",
       "      <td>CGAACATCATTGGTAC-1.05_D01</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>WMB-10Xv2-Isocortex-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3752900</th>\n",
       "      <td>AGGCCACTCTCAACTT-081_C01</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>WMB-10Xv2-OLF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3597662</th>\n",
       "      <td>GCTGCTTAGATCGGGT-1.08_B01</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>WMB-10Xv2-Isocortex-4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>682049</th>\n",
       "      <td>AGATGCTTCACCGGGT-274_A07</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>WMB-10Xv3-HPF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2112469</th>\n",
       "      <td>TCATGAGCAGCTGAAG-412_B07</td>\n",
       "      <td>WMB-10Xv3</td>\n",
       "      <td>WMB-10Xv3-HPF</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3656614</th>\n",
       "      <td>TGGTTAGAGCCAGGAT-033_B01</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>WMB-10Xv2-Isocortex-4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        cell_label dataset_label   feature_matrix_label\n",
       "1889336   GTTACAGCAGCAATTC-294_A05     WMB-10Xv3           WMB-10Xv3-TH\n",
       "1960181   AATCACGAGTGGATTA-448_A07     WMB-10Xv3  WMB-10Xv3-Isocortex-2\n",
       "3813712   GTGCATAGTCTGGTCG-069_C01     WMB-10Xv2  WMB-10Xv2-Isocortex-4\n",
       "3742959   TTCTTAGAGAACAATC-092_A01     WMB-10Xv2          WMB-10Xv2-OLF\n",
       "3407039  CGAACATCATTGGTAC-1.05_D01     WMB-10Xv2  WMB-10Xv2-Isocortex-3\n",
       "3752900   AGGCCACTCTCAACTT-081_C01     WMB-10Xv2          WMB-10Xv2-OLF\n",
       "3597662  GCTGCTTAGATCGGGT-1.08_B01     WMB-10Xv2  WMB-10Xv2-Isocortex-4\n",
       "682049    AGATGCTTCACCGGGT-274_A07     WMB-10Xv3          WMB-10Xv3-HPF\n",
       "2112469   TCATGAGCAGCTGAAG-412_B07     WMB-10Xv3          WMB-10Xv3-HPF\n",
       "3656614   TGGTTAGAGCCAGGAT-033_B01     WMB-10Xv2  WMB-10Xv2-Isocortex-4"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wmb_metadata.sample(n=10, random_state=np.random.default_rng(21231))[['cell_label', 'dataset_label', 'feature_matrix_label']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95428eaf-e913-4427-bb64-cc0a370c78f6",
   "metadata": {},
   "source": [
    "Let's download a small file from the Whole Mouse Brain data release, this one corresponding to\n",
    "- `dataset_label == WMB-10Xv2`\n",
    "- `feature_matrix_label == WMB-10Xv2-TH`\n",
    "\n",
    "**Note:** that you must specify the normalization (`raw` or `log2`) you want when you ask your `abc_cache` to download an h5ad file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0dcd6c91-7c92-4255-83dc-e4c3637ee40c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WMB-10Xv2-TH-log2.h5ad: 100%|█████████████| 4.04G/4.04G [09:40<00:00, 6.95MMB/s]\n"
     ]
    }
   ],
   "source": [
    "TH_h5ad_path = abc_cache.get_file_path(\n",
    "    directory='WMB-10Xv2',\n",
    "    file_name='WMB-10Xv2-TH/log2'\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cb07fec-ac19-466e-87b0-1b6d80ecccd2",
   "metadata": {},
   "source": [
    "Let's subset the cell_metdata to only include cells in this `dataset_label`, `feature_matrix_label` combination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "67c9c48f-b4ab-46d0-adba-7b9da43fc609",
   "metadata": {},
   "outputs": [],
   "source": [
    "TH_metadata = wmb_metadata[\n",
    "    np.logical_and(\n",
    "        wmb_metadata.dataset_label=='WMB-10Xv2',\n",
    "        wmb_metadata.feature_matrix_label=='WMB-10Xv2-TH'\n",
    "    )\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "7a6a2f85-7119-4f8f-8120-2fc6cb50b8e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "130555"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(TH_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "ccb7c19c-7c0d-4877-b229-ccf925510714",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_label</th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>barcoded_cell_sample_label</th>\n",
       "      <th>library_label</th>\n",
       "      <th>feature_matrix_label</th>\n",
       "      <th>entity</th>\n",
       "      <th>brain_section_label</th>\n",
       "      <th>library_method</th>\n",
       "      <th>region_of_interest_acronym</th>\n",
       "      <th>donor_label</th>\n",
       "      <th>donor_genotype</th>\n",
       "      <th>donor_sex</th>\n",
       "      <th>dataset_label</th>\n",
       "      <th>x</th>\n",
       "      <th>y</th>\n",
       "      <th>cluster_alias</th>\n",
       "      <th>abc_sample_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2342830</th>\n",
       "      <td>CAGGTGCAGGCTAGCA-040_C01</td>\n",
       "      <td>CAGGTGCAGGCTAGCA</td>\n",
       "      <td>040_C01</td>\n",
       "      <td>L8TX_180815_01_E08</td>\n",
       "      <td>WMB-10Xv2-TH</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv2</td>\n",
       "      <td>TH</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-404124</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>0.904733</td>\n",
       "      <td>8.810323</td>\n",
       "      <td>1000</td>\n",
       "      <td>da8ad858-d33c-49eb-aae7-df16eba116a0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2343540</th>\n",
       "      <td>TGCGCAGGTTGCGCAC-045_C01</td>\n",
       "      <td>TGCGCAGGTTGCGCAC</td>\n",
       "      <td>045_C01</td>\n",
       "      <td>L8TX_180829_01_C10</td>\n",
       "      <td>WMB-10Xv2-TH</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv2</td>\n",
       "      <td>TH</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-407905</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>F</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>-1.702466</td>\n",
       "      <td>7.063231</td>\n",
       "      <td>1746</td>\n",
       "      <td>c517bbef-bc4f-4317-8f5c-09ce80605bfe</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2343542</th>\n",
       "      <td>CGATGTATCTTGCCGT-042_B01</td>\n",
       "      <td>CGATGTATCTTGCCGT</td>\n",
       "      <td>042_B01</td>\n",
       "      <td>L8TX_180829_01_B09</td>\n",
       "      <td>WMB-10Xv2-TH</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv2</td>\n",
       "      <td>TH</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-407901</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>-10.221148</td>\n",
       "      <td>3.479807</td>\n",
       "      <td>3497</td>\n",
       "      <td>d2a78ca7-cee4-4ec9-9385-54048f6d95c6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2345608</th>\n",
       "      <td>GACTAACGTCCTCTTG-040_B01</td>\n",
       "      <td>GACTAACGTCCTCTTG</td>\n",
       "      <td>040_B01</td>\n",
       "      <td>L8TX_180815_01_D08</td>\n",
       "      <td>WMB-10Xv2-TH</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv2</td>\n",
       "      <td>TH</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-404124</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>-0.080485</td>\n",
       "      <td>7.137248</td>\n",
       "      <td>1039</td>\n",
       "      <td>aa98d92b-fe2b-4c84-8d42-b2c2d0b8b2fb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2345609</th>\n",
       "      <td>GATCGTACAACTGCTA-040_B01</td>\n",
       "      <td>GATCGTACAACTGCTA</td>\n",
       "      <td>040_B01</td>\n",
       "      <td>L8TX_180815_01_D08</td>\n",
       "      <td>WMB-10Xv2-TH</td>\n",
       "      <td>cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10Xv2</td>\n",
       "      <td>TH</td>\n",
       "      <td>Snap25-IRES2-Cre;Ai14-404124</td>\n",
       "      <td>Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt</td>\n",
       "      <td>M</td>\n",
       "      <td>WMB-10Xv2</td>\n",
       "      <td>-0.154540</td>\n",
       "      <td>7.089330</td>\n",
       "      <td>1039</td>\n",
       "      <td>f29decdc-df2c-4298-853e-c726e5fc226e</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       cell_label      cell_barcode  \\\n",
       "2342830  CAGGTGCAGGCTAGCA-040_C01  CAGGTGCAGGCTAGCA   \n",
       "2343540  TGCGCAGGTTGCGCAC-045_C01  TGCGCAGGTTGCGCAC   \n",
       "2343542  CGATGTATCTTGCCGT-042_B01  CGATGTATCTTGCCGT   \n",
       "2345608  GACTAACGTCCTCTTG-040_B01  GACTAACGTCCTCTTG   \n",
       "2345609  GATCGTACAACTGCTA-040_B01  GATCGTACAACTGCTA   \n",
       "\n",
       "        barcoded_cell_sample_label       library_label feature_matrix_label  \\\n",
       "2342830                    040_C01  L8TX_180815_01_E08         WMB-10Xv2-TH   \n",
       "2343540                    045_C01  L8TX_180829_01_C10         WMB-10Xv2-TH   \n",
       "2343542                    042_B01  L8TX_180829_01_B09         WMB-10Xv2-TH   \n",
       "2345608                    040_B01  L8TX_180815_01_D08         WMB-10Xv2-TH   \n",
       "2345609                    040_B01  L8TX_180815_01_D08         WMB-10Xv2-TH   \n",
       "\n",
       "        entity  brain_section_label library_method region_of_interest_acronym  \\\n",
       "2342830   cell                  NaN          10Xv2                         TH   \n",
       "2343540   cell                  NaN          10Xv2                         TH   \n",
       "2343542   cell                  NaN          10Xv2                         TH   \n",
       "2345608   cell                  NaN          10Xv2                         TH   \n",
       "2345609   cell                  NaN          10Xv2                         TH   \n",
       "\n",
       "                          donor_label                        donor_genotype  \\\n",
       "2342830  Snap25-IRES2-Cre;Ai14-404124  Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt   \n",
       "2343540  Snap25-IRES2-Cre;Ai14-407905  Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt   \n",
       "2343542  Snap25-IRES2-Cre;Ai14-407901  Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt   \n",
       "2345608  Snap25-IRES2-Cre;Ai14-404124  Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt   \n",
       "2345609  Snap25-IRES2-Cre;Ai14-404124  Snap25-IRES2-Cre/wt;Ai14(RCL-tdT)/wt   \n",
       "\n",
       "        donor_sex dataset_label          x         y  cluster_alias  \\\n",
       "2342830         M     WMB-10Xv2   0.904733  8.810323           1000   \n",
       "2343540         F     WMB-10Xv2  -1.702466  7.063231           1746   \n",
       "2343542         M     WMB-10Xv2 -10.221148  3.479807           3497   \n",
       "2345608         M     WMB-10Xv2  -0.080485  7.137248           1039   \n",
       "2345609         M     WMB-10Xv2  -0.154540  7.089330           1039   \n",
       "\n",
       "                                abc_sample_id  \n",
       "2342830  da8ad858-d33c-49eb-aae7-df16eba116a0  \n",
       "2343540  c517bbef-bc4f-4317-8f5c-09ce80605bfe  \n",
       "2343542  d2a78ca7-cee4-4ec9-9385-54048f6d95c6  \n",
       "2345608  aa98d92b-fe2b-4c84-8d42-b2c2d0b8b2fb  \n",
       "2345609  f29decdc-df2c-4298-853e-c726e5fc226e  "
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_metadata.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b66a0ad7-4363-43bf-ae03-6662dbc8ac86",
   "metadata": {},
   "source": [
    "Now, let's open the h5ad file we downloaded.\n",
    "\n",
    "**Note:** it is very important to open the file with `backed = 'r'`. If you do not, anndata will read the entire h5ad file into memory which, depending on the dataset, may be a lot of data that you do not want. Opening the file with `backed = 'r'` tells anndata not to load data into memory until you actually want to do something with it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "1b576f7d-0393-4510-aabb-e6a9e2ba6a2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "TH_h5ad_data = anndata.read_h5ad(\n",
    "    TH_h5ad_path,\n",
    "    backed='r'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "0fed00e0-b4bf-4a02-978c-d67a514421b9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "anndata._core.anndata.AnnData"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "type(TH_h5ad_data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "caca9243-f392-4754-9d2f-c0e12b30256d",
   "metadata": {},
   "source": [
    "We have now instantiated an `AnnData` object. `AnnData` objects can contain many member objects for containing data and its related metadata. Consult the official anndata documentation (listed at the top of this notebook) for a full list. The h5ad files distributed with `abc_atlas_access` are very stripped down, they contain only\n",
    "- `X` the array containing the cell-by-gene data. Each row in this array is a cell. Each column in this array is a gene)\n",
    "- `obs` a dataframe containing metadata related to every cell in the cell-by-gene dataset (the rows in `X`)\n",
    "- `var` a dataframe containing metadata related to every gene in the cell-by-gene dataset (the columns in `X`)\n",
    "\n",
    "Furthermore, because `abc_atlas_access` distributes metadata in many normalized CSV tables, the metadata contained in `obs` and `var` is generally the bare minimum necessary to link the rows and columns in the h5ad file to the other CSV tables in the `abc_atlas_access` dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "6079e2fb-4ca8-4e63-8876-586072f9248d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32285"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(TH_h5ad_data.var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "1f18feb4-632e-4c62-82e1-1ec81b4c4289",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_identifier</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000051951</th>\n",
       "      <td>Xkr4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000089699</th>\n",
       "      <td>Gm1992</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102331</th>\n",
       "      <td>Gm19938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000102343</th>\n",
       "      <td>Gm37381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000025900</th>\n",
       "      <td>Rp1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_symbol\n",
       "gene_identifier               \n",
       "ENSMUSG00000051951        Xkr4\n",
       "ENSMUSG00000089699      Gm1992\n",
       "ENSMUSG00000102331     Gm19938\n",
       "ENSMUSG00000102343     Gm37381\n",
       "ENSMUSG00000025900         Rp1"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_h5ad_data.var.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "45f5bc6b-6deb-4467-b67a-ef5cc0fe9b41",
   "metadata": {},
   "source": [
    "Here we see that `var` just lists the gene identifier and gene symbol of every column in `X`. Note that the `var` dataframe is indexed on the gene identifier, so that you can access individual rows in `var` using the gene identifier, i.e."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "b9cdb242-a5b6-4675-9d66-8a5e930f12a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene_symbol    Gm1992\n",
       "Name: ENSMUSG00000089699, dtype: object"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_h5ad_data.var.loc['ENSMUSG00000089699']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6411bb5d-054e-439d-9ba3-784fc6c6c725",
   "metadata": {},
   "source": [
    "Similarly, `obs` lists minimal metadata on each cell in `X`, indexed on the `cell_label`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "3aa4c75b-fa0b-4745-b28b-64e3b84c95fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>library_label</th>\n",
       "      <th>anatomical_division_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CAGGTGCAGGCTAGCA-040_C01</th>\n",
       "      <td>CAGGTGCAGGCTAGCA</td>\n",
       "      <td>L8TX_180815_01_E08</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGCGCAGGTTGCGCAC-045_C01</th>\n",
       "      <td>TGCGCAGGTTGCGCAC</td>\n",
       "      <td>L8TX_180829_01_C10</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGATGTATCTTGCCGT-042_B01</th>\n",
       "      <td>CGATGTATCTTGCCGT</td>\n",
       "      <td>L8TX_180829_01_B09</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GACTAACGTCCTCTTG-040_B01</th>\n",
       "      <td>GACTAACGTCCTCTTG</td>\n",
       "      <td>L8TX_180815_01_D08</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GATCGTACAACTGCTA-040_B01</th>\n",
       "      <td>GATCGTACAACTGCTA</td>\n",
       "      <td>L8TX_180815_01_D08</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              cell_barcode       library_label  \\\n",
       "cell_label                                                       \n",
       "CAGGTGCAGGCTAGCA-040_C01  CAGGTGCAGGCTAGCA  L8TX_180815_01_E08   \n",
       "TGCGCAGGTTGCGCAC-045_C01  TGCGCAGGTTGCGCAC  L8TX_180829_01_C10   \n",
       "CGATGTATCTTGCCGT-042_B01  CGATGTATCTTGCCGT  L8TX_180829_01_B09   \n",
       "GACTAACGTCCTCTTG-040_B01  GACTAACGTCCTCTTG  L8TX_180815_01_D08   \n",
       "GATCGTACAACTGCTA-040_B01  GATCGTACAACTGCTA  L8TX_180815_01_D08   \n",
       "\n",
       "                         anatomical_division_label  \n",
       "cell_label                                          \n",
       "CAGGTGCAGGCTAGCA-040_C01                        TH  \n",
       "TGCGCAGGTTGCGCAC-045_C01                        TH  \n",
       "CGATGTATCTTGCCGT-042_B01                        TH  \n",
       "GACTAACGTCCTCTTG-040_B01                        TH  \n",
       "GATCGTACAACTGCTA-040_B01                        TH  "
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_h5ad_data.obs.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92ae3526-244b-4029-8808-bab994638a8f",
   "metadata": {},
   "source": [
    "These `cell_labels` correspond to the `cell_labels` in `cell_metadata` that were annotated as belonging to this (`dataset_label`, `feature_matrix_label` pair).\n",
    "\n",
    "**Note:** There are more cells in the h5ad file than in the subset of `cell_metadata`. The \"extra\" cells in the h5ad file represent cells that failed QC for some reason."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "f9ae7120-17d0-4640-85f1-60de99badf5e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "130555 cells in TH_metadata\n",
      "131212 cells in this h5ad file\n",
      "0 cells TH_metadata do not appear in this h5ad file\n"
     ]
    }
   ],
   "source": [
    "print(f'{len(TH_metadata)} cells in TH_metadata')\n",
    "print(f'{len(TH_h5ad_data.obs)} cells in this h5ad file')\n",
    "delta = len(set(TH_metadata.cell_label)-set(TH_h5ad_data.obs.index.values))\n",
    "print(f'{delta} cells TH_metadata do not appear in this h5ad file')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9fde4741-6f6a-4ca6-9bec-6abdeff7e9b6",
   "metadata": {},
   "source": [
    "## Slicing real abc_atlas_access data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "572945be-e19c-47c1-898d-b2f5ec0f0104",
   "metadata": {},
   "source": [
    "With one exception, all of the data distributed by `abc_atlas_access` is stored in CSR format. This file is not that exception."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "18a7d85c-9e95-4211-be1a-19a3b86e4ed4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CSRDataset: backend hdf5, shape (131212, 32285), data_dtype float32"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_h5ad_data.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "1f426242-964f-41ce-9b54-7a37d6519999",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(131212, 32285)"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_h5ad_data.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "faacf047-e4d0-4902-bb9f-f23c1a811dec",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Naively, this data contains 4236179420 (cell, gene) pairs\n",
      "If each were represented by a 4 byte int/float, that would require 15.780998095870018 GB\n"
     ]
    }
   ],
   "source": [
    "n_cell_gene_pairs = (TH_h5ad_data.X.shape[0]*TH_h5ad_data.X.shape[1])\n",
    "gb = 4*n_cell_gene_pairs/(1024**3)\n",
    "print(f\"Naively, this data contains {n_cell_gene_pairs} (cell, gene) pairs\")\n",
    "print(f\"If each were represented by a 4 byte int/float, that would require {gb} GB\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dfd7601-ba0d-492f-8b5c-a0d20afcc294",
   "metadata": {},
   "source": [
    "Because the data is in CSR format, whatever manipulation we want to perform should be approached first as a row-slicing operation.\n",
    "\n",
    "Grabbing a subset of cells can naively be achieved as in our cartoon example above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "2938335f-5429-47e8-b6c2-ba10ea0f2374",
   "metadata": {},
   "outputs": [],
   "source": [
    "TH_cell_idx = np.array([5, 6, 7, 9001, 9002, 9003])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "d1fbf58f-0b57-48c5-a938-5ee7847f5948",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 3.13 ms, sys: 2.35 ms, total: 5.48 ms\n",
      "Wall time: 5.46 ms\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "TH_cell_subset = TH_h5ad_data[TH_cell_idx, :].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "740d9618-6f6e-4d23-8719-e640884c4ccd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_barcode</th>\n",
       "      <th>library_label</th>\n",
       "      <th>anatomical_division_label</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_label</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GTGCAGCTCACAGGCC-040_C01</th>\n",
       "      <td>GTGCAGCTCACAGGCC</td>\n",
       "      <td>L8TX_180815_01_E08</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGATCGGTCCTTTCGG-045_A01</th>\n",
       "      <td>CGATCGGTCCTTTCGG</td>\n",
       "      <td>L8TX_180829_01_F09</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TGAGCCGTCGGTGTTA-043_A01</th>\n",
       "      <td>TGAGCCGTCGGTGTTA</td>\n",
       "      <td>L8TX_180829_01_C09</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AGAATAGAGATCCCAT-037_B01</th>\n",
       "      <td>AGAATAGAGATCCCAT</td>\n",
       "      <td>L8TX_180815_01_F07</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGAGAAGAGTCCATAC-037_B01</th>\n",
       "      <td>CGAGAAGAGTCCATAC</td>\n",
       "      <td>L8TX_180815_01_F07</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CGAGAAGTCCCAAGAT-037_B01</th>\n",
       "      <td>CGAGAAGTCCCAAGAT</td>\n",
       "      <td>L8TX_180815_01_F07</td>\n",
       "      <td>TH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              cell_barcode       library_label  \\\n",
       "cell_label                                                       \n",
       "GTGCAGCTCACAGGCC-040_C01  GTGCAGCTCACAGGCC  L8TX_180815_01_E08   \n",
       "CGATCGGTCCTTTCGG-045_A01  CGATCGGTCCTTTCGG  L8TX_180829_01_F09   \n",
       "TGAGCCGTCGGTGTTA-043_A01  TGAGCCGTCGGTGTTA  L8TX_180829_01_C09   \n",
       "AGAATAGAGATCCCAT-037_B01  AGAATAGAGATCCCAT  L8TX_180815_01_F07   \n",
       "CGAGAAGAGTCCATAC-037_B01  CGAGAAGAGTCCATAC  L8TX_180815_01_F07   \n",
       "CGAGAAGTCCCAAGAT-037_B01  CGAGAAGTCCCAAGAT  L8TX_180815_01_F07   \n",
       "\n",
       "                         anatomical_division_label  \n",
       "cell_label                                          \n",
       "GTGCAGCTCACAGGCC-040_C01                        TH  \n",
       "CGATCGGTCCTTTCGG-045_A01                        TH  \n",
       "TGAGCCGTCGGTGTTA-043_A01                        TH  \n",
       "AGAATAGAGATCCCAT-037_B01                        TH  \n",
       "CGAGAAGAGTCCATAC-037_B01                        TH  \n",
       "CGAGAAGTCCCAAGAT-037_B01                        TH  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_cell_subset.obs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f59ee1cd-1865-4f42-a023-aabd05741ff8",
   "metadata": {},
   "source": [
    "We now have a subset of 6 cells drawn from our original dataset. The X matrix is still in sparse format"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "7a0448bc-d521-4b31-ae92-e8a825253c06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Compressed Sparse Row sparse matrix of dtype 'float32'\n",
       "\twith 20928 stored elements and shape (6, 32285)>"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_cell_subset.X"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1eb7c566-bb88-48a7-8202-394d9590e089",
   "metadata": {},
   "source": [
    "We can get a dense array representation of the `X` data using the `toarray()` function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "ec8e6df3-aae1-4083-b7e3-dfdb29b93658",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[8.742955 , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       [7.2100267, 0.       , 7.2100267, ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       [8.218743 , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       [8.806027 , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        0.       ],\n",
       "       [7.01022  , 0.       , 0.       , ..., 0.       , 0.       ,\n",
       "        8.004613 ]], shape=(6, 32285), dtype=float32)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_cell_subset.X.toarray()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee1479dc-101f-4f6e-8c5a-8c155dc04f2a",
   "metadata": {},
   "source": [
    "## Slicing real data by genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b95d57db-6dfb-443b-8a1d-99b8b3c4c973",
   "metadata": {},
   "source": [
    "To slice the real data by genes, it is most efficient to chunk the data by cells, slice each chunk by genes, and then concatenate the resutling `AnnData` objects into a single object. We will show you how to do this below. First, we draw your attatntion to the `chunked_X` which provides you an iterator over chhunks of rows in an `AnnData` object.\n",
    "\n",
    "**Note:** This is also a `chunk_X` function which returns a single chunk of the `AnnData` object. Be sure you are calling the correct one."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "8113f42d-043e-4ee3-9225-335d433955a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "rows_at_a_time = 10000\n",
    "row_iterator = TH_h5ad_data.chunked_X(rows_at_a_time)\n",
    "\n",
    "chunk = next(row_iterator)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "830f4771-57fc-40b8-8370-62fa9736219c",
   "metadata": {},
   "source": [
    "The chunks used by this iterator contain\n",
    "- A chunk of the `X` matrix (in sparse or dense representation depending on the form of the unchunked data)\n",
    "- The indices of the rows included in the chunk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "2d728132-a706-4da3-a7cb-90de2845a18c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(<Compressed Sparse Row sparse matrix of dtype 'float32'\n",
      "\twith 37605474 stored elements and shape (10000, 32285)>, 0, 10000)\n"
     ]
    }
   ],
   "source": [
    "print(chunk)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17766e10-e95a-4251-9789-426a9e206213",
   "metadata": {},
   "source": [
    "As said before, this can be used to assemble, one chunk at a time, a single `AnnData` object that contains a subset of genes from the original `AnnData` object.\n",
    "\n",
    "Let's select a random subset of 200 genes that we want to subset from our data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "a4a6c204-9d12-4a76-9ebc-b283a4c77f6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "rng = np.random.default_rng(871231)\n",
    "gene_idx = np.sort(\n",
    "    rng.choice(\n",
    "        np.arange(TH_h5ad_data.shape[1]),\n",
    "        200,\n",
    "        replace=False\n",
    "    )\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "ad6f475a-6987-4693-bb8c-76c21ee6cbcf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 1.29 s, sys: 595 ms, total: 1.88 s\n",
      "Wall time: 1.89 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# because we are slicing on genes,\n",
    "# each chunk in the new AnnData file will have the same\n",
    "# var dataframe\n",
    "var = TH_h5ad_data.var.iloc[gene_idx]\n",
    "\n",
    "rows_at_a_time = 1000\n",
    "row_iterator = TH_h5ad_data.chunked_X(rows_at_a_time)\n",
    "\n",
    "# loop over the chunks returned by the row_iterator, creating\n",
    "# a list of smaller AnnData objects that can later be concatenated\n",
    "# into one larger AnnData object\n",
    "sub_adata = []\n",
    "for chunk in row_iterator:\n",
    "    obs_chunk = TH_h5ad_data.obs[chunk[1]: chunk[2]]\n",
    "    x_chunk = chunk[0][:, gene_idx]\n",
    "    sub_adata.append(\n",
    "        anndata.AnnData(\n",
    "            obs=obs_chunk,\n",
    "            var=var,\n",
    "            X=x_chunk\n",
    "        )\n",
    "    )\n",
    "\n",
    "# concatenate the smaller AnnData objects into a single object\n",
    "TH_gene_subset = anndata.concat(sub_adata, merge=\"same\")\n",
    "del sub_adata\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "12b3384d-109d-4142-850e-f08670a1b41b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_symbol</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_identifier</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000037447</th>\n",
       "      <td>Arid5a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000026110</th>\n",
       "      <td>Mgat4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000026043</th>\n",
       "      <td>Col3a1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000091283</th>\n",
       "      <td>Gm17234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000026027</th>\n",
       "      <td>Stradb</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000091736</th>\n",
       "      <td>Yy2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000103347</th>\n",
       "      <td>Gm37734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000101809</th>\n",
       "      <td>Gm28852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000101566</th>\n",
       "      <td>Gm28276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUSG00000100537</th>\n",
       "      <td>Gm28485</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>200 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   gene_symbol\n",
       "gene_identifier               \n",
       "ENSMUSG00000037447      Arid5a\n",
       "ENSMUSG00000026110      Mgat4a\n",
       "ENSMUSG00000026043      Col3a1\n",
       "ENSMUSG00000091283     Gm17234\n",
       "ENSMUSG00000026027      Stradb\n",
       "...                        ...\n",
       "ENSMUSG00000091736         Yy2\n",
       "ENSMUSG00000103347     Gm37734\n",
       "ENSMUSG00000101809     Gm28852\n",
       "ENSMUSG00000101566     Gm28276\n",
       "ENSMUSG00000100537     Gm28485\n",
       "\n",
       "[200 rows x 1 columns]"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TH_gene_subset.var"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f19a5477-20c1-4580-b056-96e7a826ad65",
   "metadata": {},
   "source": [
    "Let's compare this to directly slicing the `AnnData` object all at once."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "0d55b827-f560-4762-a1b9-b9bd52834fcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 734 ms, sys: 1.3 s, total: 2.03 s\n",
      "Wall time: 2.17 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "alt = TH_h5ad_data[:, gene_idx].to_memory()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "8c7ddaa4-27b5-4d42-b8a0-9739a5a62995",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.testing.assert_array_equal(alt.X.toarray(), TH_gene_subset.X.toarray())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "0e880d6b-b73d-424e-937a-a117ef59b81e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt.obs.equals(TH_gene_subset.obs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "bde5238d-fcda-44c9-a6fa-db1885e9aaf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt.var.equals(TH_gene_subset.var)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "098f62bb-60f9-413a-8ec7-fcc5937d9c5a",
   "metadata": {},
   "source": [
    "Directly subsetting the `AnnData` object was actually a littel faster than our \"iterate over chunks\" strategy and ended up producing equivalent output. However, let us consider the maximum memory footprint of the operation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "886a54e1-61cc-429c-8290-1aef8b5d8ecd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slicing WMB-10Xv2-TH-log2.h5ad along axis 'gene'\n",
      "took 1.91e+00 s\n",
      "maximum memory footprint 7.90e-01 GB (file size is 3.76e+00 GB)\n"
     ]
    }
   ],
   "source": [
    "read_and_profile.read_by_chunks(\n",
    "    path=TH_h5ad_path,\n",
    "    gene_idx=gene_idx,\n",
    "    rows_at_a_time=rows_at_a_time\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "431f4993-87b2-423a-9d17-5ed8e0799bda",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "slicing WMB-10Xv2-TH-log2.h5ad along 'gene'\n",
      "took 1.87e+03 ms\n",
      "maximum memory footprint 3.99e+00 GB (file size is 3.76e+00 GB)\n"
     ]
    }
   ],
   "source": [
    "read_and_profile.read(\n",
    "    path=TH_h5ad_path,\n",
    "    axis='gene',\n",
    "    idx=gene_idx\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "adad61db-1009-441a-94bf-2bc139cd85ee",
   "metadata": {},
   "source": [
    "Again: directly subsetting a CSR `AnnData` object by genes required reading the entire file into memory. That was not a problem in this case. We intentionally downloaded one of the smallest h5ad files in `abc_atlas_access`. Some of the h5ad files distributed by `abc_atlas_access`, however, can be several tens of GB in size. Reading these into memory could be prohibitive. Intelligently chunking over the data to achieve your analysis goals will save you some unfortunately timed \"Out of Memory\" errors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9bf9703-2459-4f83-afc4-d28e08941940",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
